ssl/ssl_lib.c

```diff
@@ -4607,7 +4607,8 @@ void ssl_set_masks(SSL_CONNECTION *s)
 
     dh_tmp = (c->dh_tmp != NULL
               || c->dh_tmp_cb != NULL
-              || c->dh_tmp_auto);
+              || c->dh_tmp_auto
+              || tls1_shared_group(s, -2, 1, 0) != 0);
```

Comment:
The magic `-2` needs some explanation, or at least a clear macro name to signal what it means.
Even the `1` and `0` could be macros `USE_FFDHE` and `NO_EC` or better names, or perhaps:
`1 /*ffdhe*/, 0 /*ec*/`